initial trace module

tsloughter opened the pull request:
Thoughts?

tsloughter committed:
initial trace module

hauleth commented on lib/opencensus/trace.ex:
```diff
@@ -0,0 +1,27 @@
+      try do
+        unquote(block)
+      after
+        :ocp.finish_span()
```
Use `:oc_trace.finish_span/1` to be sure that the span you opened on L5 is the same that you close there.

tsloughter replied:
Ah yea, I'll fix this up.

hauleth commented on lib/opencensus/trace.ex:
```diff
@@ -0,0 +1,27 @@
+defmodule Opencensus.Trace do
+  @doc "Wraps the given block in a tracing child span with the given label/name"
+  defmacro with_child_span(label, do: block) do
```
I would probably use function instead of macro.

hauleth commented on lib/opencensus/trace.ex:
```diff
@@ -0,0 +1,27 @@
+  end
+
+  @doc "Wraps the given block in a tracing child span with the given label/name and additional attributes"
+  defmacro with_child_span(label, attributes, do: block) do
```
Both cases can be squashed into single definition by using:

```elixir
defmacro with_child_span(label, attributes \\ quote do: %{}, do: block) do
  # …
```

tsloughter replied:
That doesn't compile so I'm keeping it as two separate definitions for now.

hauleth replied:
@tsloughter `quote(do: [])`, alternatively you can do:

```elixir
defmacro with_child_span(label, do: block),
  do: with_child_span(label, quote(do: []), do: block)
```

hauleth reviewed (changes requested):
I would be wary of using macros and unless you wan to do anything in compile time then I would avoid them.

tsloughter commented:
I like the macro, and usually I hate macros.

hauleth commented:
But there is no point in using macro except making it slightly "nicer" by using `do … end` syntax. For me it is pointless and potentially can cause weird error reports. It would make sense if you would provide context as the attributes. That is why I have created census-instrumentation/opencensus-erlang#112, in that case it would make sense ;)

tsloughter committed:
use explicit span context when finishing

hauleth committed:
feat: provide additional context for span

It adds additional context to the span, namely:

- module
- line
- function where span started
- file

tsloughter committed:
use optional arguments to combine the macros

tsloughter committed:
Merge pull request #1 from hauleth/additional-context

Provide additional context for span
